Uninstrusive implementation of LIVE VIEW tables

vzakaznikov opened the pull request:
I hereby agree to the terms of the CLA available at https://yandex.ru/legal/cla/?lang=en.

--

This pull request is an unintrusive implementation of LIVE VIEW tables that were originally proposed in https://github.com/yandex/ClickHouse/pull/3925. Note that currently only LIVE VIEW tables are supported.

CREATE LIVE VIEW
--------------------
Create a LIVE VIEW table.

`CREATE [TEMPORARY] LIVE VIEW [IF NOT EXISTS] [db.]name AS SELECT ...`

ALTER LIVE VIEW
------------------
Force refresh of a LIVE VIEW table to re-execute the stored query.

`ALTER LIVE VIEW [db.]name REFRESH`

SELECT * FROM [db.]name
----------------------------
A standard SELECT query is supported for LIVE VIEW tables. The SELECT can be used in combination with the `WATCH [db.]name EVENTS` query to receive the latest result using the `_version` virtual column in the WHERE condition.

WATCH [db.]name 
--------------------
A query to watch a LIVE VIEW table that provides the latest result of the stored query which includes a virtual `_version` column.

`WATCH [db.]name [LIMIT n] [FORMAT format]`

WATCH [db.]name EVENTS
----------------------------
A query to watch LIVE VIEW table events that returns `version` column.
`WATCH [db.]name [LIMIT n] [FORMAT format]`

Settings
-----------
`temporary_live_view_timeout`

Timeout in seconds after which temporary live view table is deleted if there are no active WATCH queries.

`live_view_heartbeat_interval`

Interval in seconds of inactivity after which heartbeat is sent to keep live view WATCH query connection alive.

Formats
------------

Added new **JSONEachRowWithProgress** format that is similar to **JSONEachRow**. It
provides **row** and **progress** objects. In the case of LIVE VIEW tables, the **JSONEachRowWithProgress** format provides the heartbeat functionality via **progress** objects when executing a WATCH query on a LIVE VIEW table.

Example usage:

```
client1 :) WATCH test.lv FORMAT JSONEachRowWithProgress

WATCH test.lv
FORMAT JSONEachRowWithProgress

{"row":{"sum(a)":"0","_version":"1"}}
{"progress":{"read_rows":"1","read_bytes":"16","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
{"row":{"sum(a)":"15","_version":"2"}}
{"progress":{"read_rows":"2","read_bytes":"32","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
```
and when used with an HTTP connection

```
$ wget -O- -q "http://localhost:8123/?live_view_heartbeat_interval=1&query=WATCH test.lv FORMAT JSONEachRowWithProgress"
{"progress":{"read_rows":"1","read_bytes":"16","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
{"row":{"sum(a)":"15","_version":"2"}}
{"progress":{"read_rows":"1","read_bytes":"16","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
{"progress":{"read_rows":"1","read_bytes":"16","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
{"progress":{"read_rows":"1","read_bytes":"16","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
{"progress":{"read_rows":"1","read_bytes":"16","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}}
```


Tests
-------
Sanity tests can be found at `dbms/tests/queries/0_stateless/*_live_view_*`.
The `uexpect.py` module was added to provide support for multi-client tests.

```
Running 21 stateless tests.

00973_live_view_select:                                                 [ OK ]
00972_live_view_select_1:                                               [ OK ]
00971_live_view_watch_http_heartbeat:                                   [ OK ]
00970_live_view_watch_events_http_heartbeat:                            [ OK ]
00969_live_view_watch_format_jsoneachrowwithprogress:                   [ OK ]
00968_live_view_select_format_jsoneachrowwithprogress:                  [ OK ]
00967_live_view_watch_http:                                             [ OK ]
00966_live_view_watch_events_http:                                      [ OK ]
00965_live_view_watch_heartbeat:                                        [ OK ]
00964_live_view_watch_events_heartbeat:                                 [ OK ]
00963_temporary_live_view_watch_live_timeout:                           [ OK ]
00962_temporary_live_view_watch_live:                                   [ OK ]
00961_temporary_live_view_watch:                                        [ OK ]
00960_live_view_watch_events_live:                                      [ OK ]
00959_create_temporary_live_view:                                       [ OK ]
00958_live_view_watch_live:                                             [ OK ]
00957_live_view_watch:                                                  [ OK ]
00956_live_view_watch_events:                                           [ OK ]
00955_live_view_select_with_aggregation:                                [ OK ]
00954_live_view_select_version:                                         [ OK ]
00952_live_view_create:                                                 [ OK ]

21 tests passed. 0 tests skipped.
```

Example Usage
----------------------

WATCH query

```
client1 :) CREATE TABLE test.mt (a Int32) Engine=MergeTree order by tuple();
client1 :) CREATE LIVE VIEW test.lv AS SELECT sum(a) FROM test.mt;
client1 :) INSERT INTO test.mt VALUES (1),(2),(3)

client2 :) WATCH test.lv EVENTS

WATCH test.lv EVENTS

┌─version─┐
│       1 │ 
└─────────┘
┌─version─┐
│       2 │ 
└─────────┘
↑ Progress: 2.00 rows, 16.00 B (0.03 rows/s., 1.36 B/s.)

client2 :) WATCH test.lv

WATCH test.lv

┌─sum(a)─┬─_version─┐
│      6 │        2 │
└────────┴──────────┘
┌─sum(a)─┬─_version─┐
│     21 │        3 │
└────────┴──────────┘
↓ Progress: 2.00 rows, 32.00 B (0.32 rows/s., 5.14 B/s.)
```


HTTP Interface

```
$ wget -O- -q "http://localhost:8123/?query=WATCH test.lv EVENTS"
3
4
5
6
7

$ wget -O- -q "http://localhost:8123/?query=WATCH test.lv"
81    7
96    8
111    9
126    10
141    11
```

ludv1x committed:
Fixed SummingMergeTree. [#CLICKHOUSE-2]

alexey-milovidov committed:
Update SummingSortedBlockInputStream.h

alexey-milovidov committed:
Update SummingSortedBlockInputStream.h

proller committed:
Update build_debian.sh

ludv1x committed:
Add multi index for data_parts storage. And fixed bugs. [#CLICKHOUSE-3452]

Fixed handling of obsolete parts.
Fixed conflict resolution between simultaneous PreCommitted covering parts.
Fixed memory leak caused by ordinary MergeTree parts stucked in Deleting state.
Added hidden _state column into system.parts.

proller committed:
Zlib-ng: enable zlib compat mode (this fixes log compression) (#CLICKHOUSE-3447)

proller committed:
Zlib: as submodule, fix compile options; config: add logger options: logger.flush logger.rotateOnOpen

Conflicts:
	.gitmodules
	dbms/src/Common/BackgroundSchedulePool.h

proller committed:
Remove old contrib/libzlib-ng

proller committed:
Zlib: remove unused test targets

proller committed:
Freebsd: zlib: dont use optimizations (broken asm)

proller committed:
Fix include path contrib/zlib-ng

ludv1x committed:
Small enhancements. [#CLICKHOUSE-3452]

ludv1x committed:
Update compiler version. [#CLICKHOUSE-2]

ludv1x committed:
Added missing include dirs for dynamic compilation. [#CLICKHOUSE-2]

proller committed:
zlib fix flags in headers

proller committed:
Fix compile (missing boost libs)

Commit:
Auto version update to [54318]

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
* Work in progress to redo live view implementation to be less invasive

vzakaznikov committed:
* Adding other changes to support LIVE VIEWs

vzakaznikov committed:
* Adding support for _version virtual column to LIVE VIEWs when using SELECT
* Adding support for _version column to LIVE VIEWS when using WATCH query
* Adding initial support for WATCH query on LIVE VIEWs

vzakaznikov committed:
* Adding support for CREATE TEMPORARY LIVE VIEW
* Fixing issue with setting _version virtual column

vzakaznikov committed:
* Started to add LIVE VIEW table tests
* Added uexpect.py module
* Fixed support for CREATE TEMPORARY LIVE VIEW

vzakaznikov committed:
* Disallowing writes into LIVE VIEW tables
* Adding support for watching live views via HTTPHandler
* Small fix to WriteBufferValidUTF8.cpp to propagate next() call
* Updated copyData.cpp to treat block with no rows as
  "flush block"
* Updated PushingToViewsBlockOutputStream.cpp to directly use
  LIVE VIEW output stream instead of calling its write method

vzakaznikov committed:
* Adding a NOTICE file to include the original header that
  was inside some of the files.

vzakaznikov committed:
* Adding more LIVE VIEW tests

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
Revert "Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview"

This reverts commit 160ff4d519cdebf3816694534e19a47485193f24, reversing
changes made to a946d09f1b73efa1d04d9287da9e340a94520ac3.

vzakaznikov commented:
Hi @alexey-milovidov and @alex-zaitsev, please let me know if you think these changes meet the "unintrusive" requirement. Thanks.

vzakaznikov committed:
* Renaming live_view_heartbeat_delay setting to live_view_heartbeat_interval
  and making the setting to be in seconds instead of usec
* Updated LIVE VIEW table to keep version consistent as long as the server is up
* Removed calculation of hash in writeIntoLiveView method as it is recalculated
  in the LiveViewBlockOutputStream
* Changed to using values from the local context for live_view_heartbeat_interval and
  temporary_live_view_timeout instead of the global context
* Updated LIVE VIEW tests to match the changes

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
* Adding JSONEachRowWithProgress format
  that can be used with WATCH queries when HTTP connection is used
  in this case progress events are used as a heartbeat
* Adding new tests that use JSONEachRowWithProgress format
* Adding tests for WATCH query heartbeats when HTTP connection is used
* Small fixes to uexpect.py

vzakaznikov committed:
* Fixing typos in test names

vzakaznikov committed:
* Fixing a bug in LIVE VIEW tables. Before when a table
  is created with "AS SELECT 1" the SELECT and WATCH queries would fail
* Adding and updating LIVE VIEW tests to increase coverage

vzakaznikov committed:
* Renaming 00953_live_view_select test to avoid test number conflict

vzakaznikov committed:
* Fixing clang errors
* Fixing LIVE VIEW tests

vzakaznikov committed:
* Style check fixes

vzakaznikov committed:
* Fixing bug in dbms/src/DataStreams/PushingToViewsBlockOutputStream.cpp
  that was introduced when trying to fix a clang error.

vzakaznikov committed:
* Moving uexpect.py to dbms/tests/integration/helpers folder
* Updating uexpect.py to support close() method and
  context manager for proper resource cleanup
* Updating tests to use uexpect context manager

vzakaznikov committed:
* Fixing uexpect.py close() method and forcing to close the process
  by default
* Updating tests to be more robust

vzakaznikov committed:
* Moving uexpect.py so that it can be found
* Trying to fix failed style checks
* Trying to fix clang build fail

vzakaznikov committed:
* Trying again to fix style checks. Now running ./utils/check-style/check-style reports no errors.

vzakaznikov committed:
* Fixing errors that prevented clang build.

vzakaznikov committed:
* Fixing and reorganazing LIVE VIEW tests

vzakaznikov committed:
Fixing bugs reported by sanitizers
* data race condition on storage is_dropped access
* invalid reference to storage after it is deleted
* invalid reference to mutex after storage is deleted

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
* Fixing styling check error

vzakaznikov committed:
* Increasing timeout from 2 to 20 sec.
* Style fix

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
* Updates to sync with yandex/master

vzakaznikov committed:
* Sync with yandex/master so that the git diff yandex/master is empty

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
* Reverting changes to clickhouse-test as it seems to cause some tests to fail
* Updating client.py to use CLICKHOUSE_BINARY env variable instead of CLICKHOUSE_CLIENT

vzakaznikov committed:
* Fixing a bug in StorageLiveView.cpp in the getHeader() method
  when header block would not match blocks in the stream
  when Const column is present.
* Updating 00963_temporary_live_view_watch_live_timeout.py
  and removed timeout=5 after WATCH query is aborted with Ctrl-C
  so that the default timeout of 20sec is used
* Small style fixes in tests

vzakaznikov committed:
* Fixing bug in StorageLiveView.cpp getHeader() method
  that sometimes addressed invalid pointer when getHeader()
  method was called without holding the mutex

vzakaznikov committed:
* Fixing a bug in writeIntoLiveView method where blocks->front()
  is called before checking if blocks vector is empty.

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
* Updating tests as SHOW TABLES 'lv' does not return any tables
  as test database is not selected by default.

vzakaznikov committed:
* Removing 'hash' column from WATCH [db.live_view] EVENTS query result
  as hash of the result are not usually provided in ClickHouse.
  This is also done to match the implementation of the SELECT query that only
  returns '_version' virtual column when using LIVE VIEW tables.

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov commented:
Hi @KochetovNicolai, I have addressed all the failing checks and the PR is ready for your review. The only fail left does not seem to be caused by the changes related to LIVE VIEW tables. Thanks.

vzakaznikov committed:
* Fixing bug in writeIntoLiveView method where mergeable blocks
  would always be reobtained after 64 inserts

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
* Adding max_live_view_insert_blocks_before_refresh setting
  with default value of 64
* Updating writeIntoLiveView method to use max_live_view_insert_blocks_before_refresh
  from global context instead of hard coded number
* Adding squashing of the result blocks similar how it is done in materialized views
* Fixing bug in writeIntoLiveView that caused incoming block to be processed twice
  when maximum number of insert blocks was reached

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse into liveview

KochetovNicolai commented on dbms/src/DataStreams/BlocksBlockInputStream.h:
```diff
@@ -0,0 +1,51 @@
+{
+public:
+    /// Acquires shared ownership of the blocks vector
+    BlocksBlockInputStream(std::shared_ptr<BlocksPtr> blocks_ptr_, Block header)
```
blocks_ptr_, header are copied each invocation. You can use const reference or `std::move`.

vzakaznikov replied:
Adding std::move and grabbing blocks_ptr_ content.

```cpp
    /// Acquires shared ownership of the blocks vector
    BlocksBlockInputStream(const std::shared_ptr<BlocksPtr> & blocks_ptr_, Block header)
        : blocks(*blocks_ptr_), it((*blocks_ptr_)->begin()), end((*blocks_ptr_)->end()), header(std::move(header)) {}
```

KochetovNicolai commented on dbms/src/DataStreams/BlocksBlockInputStream.h:
```diff
@@ -0,0 +1,51 @@
+    Block readImpl() override
+    {
+        if (it == end)
+            return Block();
```
Probably, if you release `blocks_ptr` at the end, it will save some memory.

vzakaznikov replied:
Do you mean something like this?

```cpp
    Block readImpl() override
    {
        if (it == end)
        {
            blocks_ptr.reset();
            return Block();
        }

        Block res = *it;
        ++it;
        return res;
    }
```

I can add it, but I didn't see other datastreams releasing any resources in readImpl(). 
I would leave it as is, and count that once BlocksBlockInputStream goes out of scope `blocks_ptr` will be released during BlocksBlockInputStream destruction.

KochetovNicolai replied:
Ok

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+    }
+    /// length default -2 because we want LIMIT to specify number of updates so that LIMIT 1 waits for 1 update
+    /// and LIMIT 0 just returns data without waiting for any updates
+    LiveViewBlockInputStream(std::shared_ptr<StorageLiveView> storage_,
```
`storage_`, `blocks_ptr_`, `blocks_metadata_ptr_`, `active_ptr_` are copied. Use `const &` or `std::move`.

vzakaznikov replied:
Adding std::move

```cpp
    LiveViewEventsBlockInputStream(std::shared_ptr<StorageLiveView> storage_,
        std::shared_ptr<BlocksPtr> blocks_ptr_,
        std::shared_ptr<BlocksMetadataPtr> blocks_metadata_ptr_,
        std::shared_ptr<bool> active_ptr_,
        const bool has_limit_, const UInt64 limit_,
        const UInt64 heartbeat_interval_sec_,
        const UInt64 temporary_live_view_timeout_sec_)
        : storage(std::move(storage_)), blocks_ptr(std::move(blocks_ptr_)), blocks_metadata_ptr(std::move(blocks_metadata_ptr_)), active_ptr(std::move(active_ptr_)), has_limit(has_limit_), limit(limit_), heartbeat_interval_usec(heartbeat_interval_sec_ * 1000000), temporary_live_view_timeout_sec(temporary_live_view_timeout_sec_)
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+/**
+  */
+
+class LiveViewBlockInputStream : public IBlockInputStream
```
Comment is needed.

vzakaznikov replied:
Adding comment

```cpp
/** Implements LIVE VIEW table WATCH EVENTS input stream.
 *  Keeps stream alive by outputing blocks with no rows
 *  based on period specified by the heartbeat interval.
 */
class LiveViewEventsBlockInputStream : public IBlockInputStream
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+    Blocks::iterator begin;
+    /// Length specifies number of updates to send, default -1 (no limit)
+    int64_t length;
+    bool end_of_blocks{0};
```
`= false;`

vzakaznikov replied:
Updating to

```cpp
    bool end_of_blocks = false;
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+    UInt64 heartbeat_interval;
+    UInt64 temporary_live_view_timeout;
+    String blocks_hash;
+    UInt64 last_event_timestamp{0};
```
`= 0;`

vzakaznikov replied:
Updating to

```cpp
    UInt64 last_event_timestamp_usec = 0;
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+        if (!storage->is_dropped && blocks_ptr.use_count() < 3)
+            storage->startNoUsersThread(temporary_live_view_timeout);
+    }
+    /// length default -2 because we want LIMIT to specify number of updates so that LIMIT 1 waits for 1 update
```
This logic is very difficult to understand (I spent several minutes couldn't).
Just separate logic into more variables. Create `bool has_limit` to check if you need to limit updates, `const UInt64 limit` to remember limit specified by user and `UInt64 num_updates` to count updates.

vzakaznikov replied:
Adding has_limit, limit and num_updates.

```cpp
    const bool has_limit;
    const UInt64 limit;
    Int64 num_updates = -1;
```
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+    void refresh()
+    {
+        if (active && blocks && it == end)
+            it =  blocks->begin();
```
2 spaces

vzakaznikov replied:
Removing extra empty space.

```cpp
    void refresh()
    {
        if (active && blocks && it == end)
            it = blocks->begin();
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+        std::shared_ptr<BlocksPtr> blocks_ptr_,
+        std::shared_ptr<BlocksMetadataPtr> blocks_metadata_ptr_,
+        std::shared_ptr<bool> active_ptr_,
+        int64_t length_, const UInt64 & heartbeat_interval_,
```
You can use `Int64 and UInt64` or `int64_t and unit64_t`. It's strange to have a mix of types.
it's not reasonable to pass heartbeat_interval_ by reference (inside it's implemented as pointer). just copy it.

vzakaznikov replied:
Removed use of mixed types.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+        std::shared_ptr<bool> active_ptr_,
+        int64_t length_, const UInt64 & heartbeat_interval_,
+        const UInt64 & temporary_live_view_timeout_)
+        : storage(storage_), blocks_ptr(blocks_ptr_), blocks_metadata_ptr(blocks_metadata_ptr_), active_ptr(active_ptr_), length(length_ + 1), heartbeat_interval(heartbeat_interval_ * 1000000), temporary_live_view_timeout(temporary_live_view_timeout_),
```
Can't understand which time unit do we use for `heartbeat_interval_`, `heartbeat_interval`, `temporary_live_view_timeout_`. And why `* 1000000`. You can add a comment or name like `heartbeat_interval_ms`.

vzakaznikov replied:
Changed variable names to use "*_sec" and "*_usec" postfix.

```cpp
const UInt64 heartbeat_interval_sec_,
        const UInt64 temporary_live_view_timeout_sec_)
        : storage(std::move(storage_)), blocks_ptr(std::move(blocks_ptr_)), blocks_metadata_ptr(std::move(blocks_metadata_ptr_)), active_ptr(std::move(active_ptr_)), has_limit(has_limit_), limit(limit_), heartbeat_interval_usec(heartbeat_interval_sec_ * 1000000), temporary_live_view_timeout_sec(temporary_live_view_timeout_sec_)
```

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,223 @@
+        int64_t length_, const UInt64 & heartbeat_interval_,
+        const UInt64 & temporary_live_view_timeout_)
+        : storage(storage_), blocks_ptr(blocks_ptr_), blocks_metadata_ptr(blocks_metadata_ptr_), active_ptr(active_ptr_), length(length_ + 1), heartbeat_interval(heartbeat_interval_ * 1000000), temporary_live_view_timeout(temporary_live_view_timeout_),
+        blocks_hash("")
```
Seems like you don't use `blocks_hash` anywhere.

vzakaznikov replied:
Removed unused blocks_bash.

vzakaznikov committed:
* Updates to fix issues after the merge

vzakaznikov committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse

vzakaznikov committed:
Merge branch 'master' of https://github.com/vzakaznikov/ClickHouse into liveview

vzakaznikov committed:
* Updates that incorporate comments from the review

vzakaznikov commented:
Hi @KochetovNicolai, thanks for the review! I have committed the changes based on your comments.

vzakaznikov committed:
* Fixing clang build by converting old style cast (Int64) to static_cast<Int64>.

vzakaznikov committed:
* Updating tests to send Ctrl-C instead of SIGINT signal to abort WATCH query
* Updating client.py to start commands inside shell
* Removing test code from uexpect.py

vzakaznikov committed:
* Updating live view tests to workaround an issue with clickhouse-client exiting when
  Ctrl-C is pressed during WATCH query execution instead of going to the
  client prompt.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    {
+        /// Start storage no users thread
+        /// if we are the last active user
+        if (!storage->is_dropped && blocks_ptr.use_count() < 3)
```
I think it's not very safe to check use_count.
If somebody else will copy blocks_ptr, no thread will be created.

Also, code is not thread-safe. Two threads which concurrently call `~LiveViewBlockInputStream()` could see that condition `blocks_ptr.use_count() < 3` is not true, and only after that remove their ptrs. So, `storage->startNoUsersThread` won't be called.

I think the good option to solve this is to use manual atomic use counter. You may create class like 
```c++
class SharedBlocks
{
private:
    BlocksPtr blocks;
    std::atomic<size_t> use_count;
}
``` 
and RAII guard class that will anatomically decrease use_count in destructor and call something if it is 0.

Or maybe use custom deleter for shared_ptr. Don't know yet if it can help.

KochetovNicolai replied:
Also, if you add `mutex` and `condition` in `SharedBlocks`, it will probably be possible not to store `storage` field.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+        const bool has_limit_, const UInt64 limit_,
+        const UInt64 heartbeat_interval_sec_,
+        const UInt64 temporary_live_view_timeout_sec_)
+        : storage(std::move(storage_)), blocks_ptr(std::move(blocks_ptr_)), blocks_metadata_ptr(std::move(blocks_metadata_ptr_)), active_ptr(std::move(active_ptr_)), has_limit(has_limit_), limit(limit_), heartbeat_interval_usec(heartbeat_interval_sec_ * 1000000), temporary_live_view_timeout_sec(temporary_live_view_timeout_sec_)
```
Please, split this string. it's very long.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    {
+        active = active_ptr.lock();
+        {
+            if (!blocks || blocks.get() != (*blocks_ptr).get())
```
It looks like we can write `blocks = (*blocks_ptr)` without checking any condition. Am I mistaken?

Also, I think we need to create `Poco::FastMutex::ScopedLock lock(storage->mutex)`. Otherwise, it's not safe to read `blocks_ptr`.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+                if (!active)
+                    return { Block(), false };
+                /// If we are done iterating over our blocks
+                /// and there are new blocks availble then get them
```
available

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+                    }
+                    while (true)
+                    {
+                        UInt64 timestamp_usec = static_cast<UInt64>(timestamp.epochMicroseconds());
```
`Timestamp` stores value that never changes. `timestamp.epochMicroseconds() ` is implemented as:
```c++
inline Timestamp::TimeVal Timestamp::epochMicroseconds() const
{
	return _ts;
}
```
You can use default constructor `Timestamp().epochMicroseconds()` and don't store `timestamp` field.

KochetovNicolai replied:
Or call `timestamp.update()` each time.

KochetovNicolai replied:
Oh, also it's even not correct to use `Timestamp` class here.
Because `Timestamp::update()` is implemented via `system_clock`:
```c++
void Timestamp::update()
{
	_ts = std::chrono::duration_cast<std::chrono::microseconds>
		(std::chrono::system_clock::now().time_since_epoch()).count();
}
```
Actually, you can copy this code and replace `system_clock` to `steady_clock`.
See https://stackoverflow.com/questions/13263277/difference-between-stdsystem-clock-and-stdsteady-clock

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+                    while (true)
+                    {
+                        UInt64 timestamp_usec = static_cast<UInt64>(timestamp.epochMicroseconds());
+                        bool signaled = storage->condition.tryWait(storage->mutex, std::max(static_cast<UInt64>(0), heartbeat_interval_usec - (timestamp_usec - last_event_timestamp_usec)) / 1000);
```
Please split this row.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    {
+        Block res;
+
+        if (has_limit && num_updates == static_cast<Int64>(limit))
```
It's much better now!
I would prefer to use `UInt64 num_updates = 0`  and check that `num_updates + 1 == limit`. But current implementation is also fine (still, need comment why num_updates is -1 at the beginning).

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+private:
+    std::shared_ptr<StorageLiveView> storage;
+    std::shared_ptr<BlocksPtr> blocks_ptr;
+    std::shared_ptr<BlocksMetadataPtr> blocks_metadata_ptr;
```
Looks like this field is not used (or just holds object, need comment then).

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    std::weak_ptr<bool> active_ptr;
+    std::shared_ptr<bool> active;
+    BlocksPtr blocks;
+    BlocksMetadataPtr blocks_metadata;
```
Is not used.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    BlocksMetadataPtr blocks_metadata;
+    Blocks::iterator it;
+    Blocks::iterator end;
+    Blocks::iterator begin;
```
I don't see if we need `begin` somewhere. Loos like it is only being assigned but never read.

KochetovNicolai replied:
Also probably it's no need to store `end`, because we can call `blocks->end()` any time.

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    const UInt64 limit;
+    Int64 num_updates = -1;
+    bool end_of_blocks = false;
+    UInt64 heartbeat_interval_usec;
```
const ?

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+    Int64 num_updates = -1;
+    bool end_of_blocks = false;
+    UInt64 heartbeat_interval_usec;
+    UInt64 temporary_live_view_timeout_sec;
```
const ?

KochetovNicolai commented on dbms/src/DataStreams/LiveViewBlockInputStream.h:
```diff
@@ -0,0 +1,222 @@
+
+    void suspend()
+    {
+        active.reset();
```
I see that active flag is from storage. And storage is in our field.
So we can just create `std::atomic_bool active` field in `StorageLiveView` and read this field in resume.

KochetovNicolai commented:
@vzakaznikov I actually went through only couple of files, because diff is big. I think it will take some time.
Anyway, implementation is not intrusive indeed, so it will likely be merged. Thank you for it.

jakesyl commented:
I'm seeing `Code: 182. DB::Exception: Received from localhost:9000. DB::Exception: Storage Blocks doesn't support PREWHERE.` even with SELECT *. Schema is laid out in detail on #11459

KochetovNicolai commented:
@vzakaznikov I think we just need to disable `prewhere` for LV. Coult you please take a look?

jakesyl commented:
By the way, I don't think I have `optimize_move_to_prewhere` on, will check this out now
